README.rst
============
Najdi-si-sms
============



.. image:: https://travis-ci.org/brodul/najdi-si-sms.svg?branch=master
        :target: https://travis-ci.org/brodul/najdi-si-sms
        :alt: Master Travis CI Status

.. image:: https://readthedocs.org/projects/najdisi-sms/badge/?version=latest
        :target: http://najdisi-sms.readthedocs.io/en/latest/
        :alt: Latest Documentation Status

Command line utility to send automated sms through Slovenian Najdi.si service (40 sms per day for free).

For additional information visit `the documentation`_

.. _`the documentation`: http://najdisi-sms.readthedocs.io/en/latest/

How to use
==========


Installation
++++++++++++

From pypi
---------

You need the python-pip package and super user access to install system wide::

  pip install najdisi-sms

You can also install the package in virtualenv::

  virtualenv venv
  source venv/bin/activate
  pip install najdisi-sms

The CLI command can be found in "venv/bin/najdisi-sms" and you can add it to you PATH or call it directly.


From source
-----------

You can install the package systemwide with (you need su access)::

  make install
  #or
  pip install .

If you dont want/have super user access, you can install it in a python virtual env
in the root folder call::

  virtualenv venv
  source venv/bin/activate
  pip install .

The CLI command can be found in "venv/bin/najdisi-sms" and you can add it to you PATH or call it directly.

Standalone CLI command
++++++++++++++++++++++

::

  Usage: najdisi-sms -u username -p password  RECEIVER_NUM  MESSAGE

  Options:
    -h, --help            show this help message and exit
    -u USERNAME, --username=USERNAME
                          Username
    -p PASSWORD, --password=PASSWORD
                          Password
    -A USERAGENT, --useragent=USERAGENT
                          HTTP User Agent

Example::

  najdisi-sms -u brodul -p FUBAR_PASS 031123456 "Pikica, rad te mam. (sent from cronjob)"

Python API
++++++++++

Example::

  from najdisi_sms import SMSSender

  sms = SMSSender('username', 'password')
  sms.send(
      '031123456',
      'Pikica, rad te mam. (sent from cronjob)'
  )


CHANGELOG.rst
Changelog
=========

1.0.1 (unreleased)
------------------

- Nothing changed yet.


1.0.0 (2016-06-26)
------------------
- Changed the API because of wrong spelling of word 'reciever'
- Upgrade BeautifulSoup
- Refactor code
- Add changelog
- Add documentation


tests/__init__.py
import os
import unittest
from unittest import TestCase

import betamax
from betamax_serializers import pretty_json
from betamax.fixtures import unittest as unitest


najdisi_username = os.environ.get('USERNAME', 'fubar_username')
najdisi_password = os.environ.get('PASSWORD', 'fubar_password')
najdisi_phonenum = os.environ.get('PHONENUM', 'fubar_phonenum')

here_ = os.path.dirname(os.path.realpath(__file__))
betamax.Betamax.register_serializer(pretty_json.PrettyJSONSerializer)
with betamax.Betamax.configure() as config:
    config.cassette_library_dir = os.path.join(here_, 'fixtures', 'cassettes')
    config.default_cassette_options['serialize_with'] = 'prettyjson'
    config.define_cassette_placeholder(
        '<NAJDISI_USERNAME>',
        najdisi_username
    )
    config.define_cassette_placeholder(
        '<NAJDISI_PASSWORD>',
        najdisi_password
    )
    config.define_cassette_placeholder(
        '<NAJDISI_PHONENUM>',
        najdisi_phonenum
    )


class VCRTestCase(unitest.BetamaxTestCase):
    # Add your the rest of the helper methods you want for your
    # integration tests
    pass


class LoadTestCase(VCRTestCase):
    """docstring for LoadTestCase"""
    def test_dummy(self):
        import najdisi_sms
        sender = najdisi_sms.SMSSender(najdisi_username, najdisi_password)
        sender.s = self.session
        sender.send(najdisi_phonenum, "test")


if __name__ == '__main__':
    unittest.main()


najdisi_sms/api.py
# -*- coding: utf-8 -*-

import requests
from bs4 import BeautifulSoup

from najdisi_sms import utils

log = utils.get_logger()


class SMSSender(object):
    """Docstring for SMSSender. """

    def __init__(self, username, password, useragent=""):
        """@todo: to be defined1. """
        self.username = username
        self.password = password
        da = "Mozilla/5.0 (Windows; U; Windows NT 6.1; es-ES; rv:1.9.2.3)" \
            + "Gecko/20100401 Firefox/3.6.3"
        self.useragent = useragent or da
        self.s = requests.Session()

        self.daily_max = None
        self.daily_left = None

    def normalize_reciever(self, reciever_num):
        """
        Split telephone number into area code and local number.


        :reciever_num: Telephone number string.
        :returns: Tuple with area code and local number.

        """
        # 031 123 456
        who = reciever_num.strip()

        # don't change
        # 031 123 456 => 123456
        recipent = who.replace(' ', '')[3:]
        # 031 123 456 =>  031
        base_code = who[:3]

        return base_code, recipent

    def check_msg_leng(self, msg):
        """
        Checks the message length raises an exception if more than 160 chars.

        :msg: Message
        :returns: Returns non modified msg

        """
        if len(msg) > 160:
            raise Exception('Message to long')

        return msg

    def send(self, reciever, msg):
        """send the message.

        :reciever: reciever number (only Slovenian supported)
        :msg: SMS body message
        :returns: True if sending succeeded, else False.

        """

        msg = self.check_msg_leng(msg)

        base_code, recipient = self.normalize_reciever(reciever)

        log.info('Network code: %s', base_code)
        log.info('reciever: %s', recipient)
        log.info('Message: %s', msg)
        log.info('Sending SMS ...')

        self.s.headers.update({'User-Agent': self.useragent})

        response = self.s.get(
            'http://www.najdi.si/najdi.layoutnajdi.loginlink:login?t:ac=sms'
        )
        assert 'Pozabil/-a sem geslo' in response.text

        soup = BeautifulSoup(response.text, "html.parser")
        formdata_els = soup.findAll(attrs={'name': 't:formdata'})
        formdata_value = formdata_els[0].attrs['value']

        data = {
            't:formdata': formdata_value,
            'jsecLogin': self.username,
            'jsecRememberMe': 'on',
            'jsecPassword': self.password
        }
        response = self.s.post(
            'http://www.najdi.si/prijava.jsecloginform',
            data
        )
        assert 'Prejemnik' in response.text
        assert u'Pošiljatelj' in response.text
        assert 'poslani SMS-i' in response.text

        soup = BeautifulSoup(response.text, 'html.parser')

        smsno_div = soup.find('div', class_='smsno')
        self.daily_left, self.daily_max = smsno_div.strong.text.split(' / ')

        formdata_els = soup.findAll(attrs={'name': 't:formdata'})
        formdata_vals = [formdata_el.attrs['value'] for
                         formdata_el in formdata_els]

        hidden_els = soup.findAll(attrs={'name': 'hidden'})
        hidden_value = hidden_els[0].attrs['value']

        data = {
            't:ac': 'sms',
            't:formdata': formdata_vals,
            'areaCodeRecipient': base_code,
            'phoneNumberRecipient': recipient,
            'selectLru': '',
            'hidden': hidden_value,
            'name': '',
            'text': msg,
            't:submit': '["send","send"]',
            't:zoneid': 'smsZone'
        }
        response = self.s.post(
            "http://www.najdi.si/"
            "najdi.shortcutplaceholder.freesmsshortcut.smsform",
            data,
            headers={"X-Requested-With": "XMLHttpRequest"}
        )

        html_response = response.json()['content']
        assert 'Zaradi varnosti' in html_response

        soup = BeautifulSoup(html_response, 'html.parser')
        sender = soup.find('div', class_='sender').strong.text
        reciever = soup.find('div', class_='reciever').strong.text
        sent_text = soup.find('div', class_='msg').strong.text
        left_today = soup.find('div', class_='msgleft').strong.text

        return_dict = {
            'sender': sender,
            'reciever': reciever,
            'text': sent_text,
            'left_today': left_today,
        }

        log.info('Sent message:\n%s', sent_text)
        log.info('Message left for today: %s', left_today)
        log.debug('%s', return_dict)

        return return_dict


najdisi_sms/cli.py
from argparse import ArgumentParser
import os
import sys

from six.moves.configparser import ConfigParser

from . import log
from .api import SMSSender


class SettingParser(object):
    """docstring for SettingParser"""
    def __init__(self, args=None):
        self.args = args or sys.argv[1:]

        home = os.path.expanduser('~')
        self.default_config_path = os.path.join(
            home,
            '.config',
            'najdisi_sms.ini'
        )

        self.argparser = self._create_argparser()
        self.parser_space = self.argparser.parse_args(self.args)

        self.namespace = self.merge_settings(self.parser_space)
        self.check_password_username(self.namespace)

    def _create_argparser(self):
        parser = ArgumentParser()
        parser.add_argument(
            "rec_num",
            metavar=u"reciever_NUM",
            help="slovenian phone number starting with 0"
        )
        parser.add_argument(
            "message",
            metavar=u"MESSAGE",
            help="SMS message (less than 160 characters)"
        )
        parser.add_argument(
            "-c",
            "--configfile",
            dest="config",
            help=u"Config file",
            default=self.default_config_path
        )
        parser.add_argument(
            "-u",
            "--username",
            dest="username",
            help=u"Username"
        )
        parser.add_argument(
            "-p",
            "--password",
            dest="password",
            help=u"Password"
        )
        parser.add_argument(
            "-A",
            "--useragent",
            dest="useragent",
            help=u"HTTP User Agent",
            default=("Mozilla/5.0 "
                     "(Windows; U; Windows NT 6.1; es-ES; rv:1.9.2.3)"
                     "Gecko/20100401 Firefox/3.6.3")
        )
        return parser

    def merge_settings(self, parser_space):
        """
        Merge config file and cli options
        """

        def parse_ini(file_path):
            config = ConfigParser()
            config.read(file_path)
            return config

        if os.path.exists(parser_space.config):
            ini_config = parse_ini(parser_space.config)
            for attr in ['username', 'password']:
                setattr(
                    parser_space,
                    attr,
                    getattr(parser_space, attr, None) or
                    ini_config.get('najdisi_sms', attr)
                )
        elif not self.default_config_path == parser_space.config:
            log.info('Config file you specified not found!')

        return parser_space

    def check_password_username(self, namespace):
        for attr in ['username', 'password']:
            if not getattr(namespace, attr):
                raise LookupError("Missing {}!".format(attr))


def main():
    parser = SettingParser()
    namespace = parser.namespace

    sender = SMSSender(
        namespace.username,
        namespace.password,
        namespace.useragent
    )
    sender.send(namespace.rec_num, namespace.message)


if __name__ == '__main__':
    main()


najdisi_sms.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""

"""
from argparse import ArgumentParser
import sys
if sys.version_info < (3, 0):
    from ConfigParser import ConfigParser
else:
    from configparser import ConfigParser
import os  # noqa
import logging  # noqa

import requests  # noqa
from bs4 import BeautifulSoup  # noqa

logging.basicConfig()
log = logging.getLogger("najdisi_sms")
log.setLevel(logging.INFO)


class SettingParser(object):
    """docstring for SettingParser"""
    def __init__(self, args=None):
        self.args = args or sys.argv[1:]

        home = os.path.expanduser('~')
        self.default_config_path = os.path.join(
            home,
            '.config',
            'najdisi_sms.ini'
        )

        self.argparser = self._create_argparser()
        self.parser_space = self.argparser.parse_args(self.args)

        self.namespace = self.merge_settings(self.parser_space)
        self.check_password_username(self.namespace)

    def _create_argparser(self):
        parser = ArgumentParser()
        parser.add_argument(
            "rec_num",
            metavar=u"RECEIVER_NUM",
            help="slovenian phone number starting with 0"
        )
        parser.add_argument(
            "message",
            metavar=u"MESSAGE",
            help="SMS message (less than 160 characters)"
        )
        parser.add_argument(
            "-c",
            "--configfile",
            dest="config",
            help=u"Config file",
            default=self.default_config_path
        )
        parser.add_argument(
            "-u",
            "--username",
            dest="username",
            help=u"Username"
        )
        parser.add_argument(
            "-p",
            "--password",
            dest="password",
            help=u"Password"
        )
        parser.add_argument(
            "-A",
            "--useragent",
            dest="useragent",
            help=u"HTTP User Agent",
            default=("Mozilla/5.0 "
                     "(Windows; U; Windows NT 6.1; es-ES; rv:1.9.2.3)"
                     "Gecko/20100401 Firefox/3.6.3")
        )
        return parser

    def merge_settings(self, parser_space):
        """
        Merge config file and cli options
        """

        def parse_ini(file_path):
            config = ConfigParser()
            config.read(file_path)
            return config

        if os.path.exists(parser_space.config):
            ini_config = parse_ini(parser_space.config)
            for attr in ['username', 'password']:
                setattr(
                    parser_space,
                    attr,
                    getattr(parser_space, attr, None) or
                    ini_config.get('najdisi_sms', attr)
                )
        elif not self.default_config_path == parser_space.config:
            log.info('Config file you specified not found!')

        return parser_space

    def check_password_username(self, namespace):
        for attr in ['username', 'password']:
            if not getattr(namespace, attr):
                raise LookupError("Missing {}!".format(attr))


def main():
    parser = SettingParser()
    namespace = parser.namespace

    sender = SMSSender(
        namespace.username,
        namespace.password,
        namespace.useragent
    )
    sender.send(namespace.rec_num, namespace.message)


class SMSSender(object):
    """Docstring for SMSSender. """

    def __init__(self, username, password, useragent=""):
        """@todo: to be defined1. """
        self.username = username
        self.password = password
        da = "Mozilla/5.0 (Windows; U; Windows NT 6.1; es-ES; rv:1.9.2.3)" \
            + "Gecko/20100401 Firefox/3.6.3"
        self.useragent = useragent or da
        self.s = requests.Session()

    def normalize_receiver(self, receiver_num):
        """
        Split telephone number into area code and local number.


        :receiver_num: Telephone number string.
        :returns: Tuple with area code and local number.

        """
        # 031 123 456
        who = receiver_num.strip()

        # don't change
        # 031 123 456 => 123456
        recipent = who.replace(' ', '')[3:]
        # 031 123 456 =>  031
        base_code = who[:3]

        return base_code, recipent

    def check_msg_leng(self, msg):
        """
        Checks the message length raises an exception if more than 160 chars.

        :msg: Message
        :returns: Returns non modified msg

        """
        if len(msg) > 160:
            raise Exception('Message to long')

        return msg

    def send(self, receiver, msg):
        """send the message.

        :receiver: Receiver number (only Slovenian supported)
        :msg: SMS body message
        :returns: True if sending succeeded, else False.

        """

        msg = self.check_msg_leng(msg)

        base_code, recipient = self.normalize_receiver(receiver)

        log.info('Network code: %s', base_code)
        log.info('Receiver : %s', recipient)
        log.info('Message: %s', msg)
        log.info('Sending SMS ...')

        self.s.headers.update({'User-Agent': self.useragent})

        response = self.s.get(
            'http://www.najdi.si/najdi.layoutnajdi.loginlink:login?t:ac=sms'
        )

        soup = BeautifulSoup(response.text, "html.parser")
        formdata_els = soup.findAll(attrs={'name': 't:formdata'})
        formdata_value = formdata_els[0].attrs['value']

        data = {
            't:formdata': formdata_value,
            'jsecLogin': self.username,
            'jsecRememberMe': 'on',
            'jsecPassword': self.password
        }
        response = self.s.post(
            'http://www.najdi.si/prijava.jsecloginform',
            data
        )

        soup = BeautifulSoup(response.text)

        formdata_els = soup.findAll(attrs={'name': 't:formdata'})
        formdata_vals = [formdata_el.attrs['value']
                         for formdata_el in formdata_els]

        hidden_els = soup.findAll(attrs={'name': 'hidden'})
        hidden_value = hidden_els[0].attrs['value']

        data = {
            't:ac': 'sms',
            't:formdata': formdata_vals,
            'areaCodeRecipient': base_code,
            'phoneNumberRecipient': recipient,
            'selectLru': '',
            'hidden': hidden_value,
            'name': '',
            'text': msg,
            't:submit': '["send","send"]',
            't:zoneid': 'smsZone'
        }
        response = self.s.post(
            "http://www.najdi.si/"
            "najdi.shortcutplaceholder.freesmsshortcut.smsform",
            data,
            headers={"X-Requested-With": "XMLHttpRequest"}
        )
        soup = BeautifulSoup(response.text)

if __name__ == '__main__':
    main()
